README.md
# Consolidate #

### About ###
Combine resources (js, css) according to regex patterns,
to have the ideal number of requests, and minimise bloat.

Parts of this plugin are based on Minify (Vanilla/Garden):
https://github.com/66Ton99/Vanilla/tree/master/plugins/Minify

### Tested on ###
2.0.18, 2.1

### Sponsor ###
Special thanks to balaboostas.com for making this happen.



changelog.txt
Fri Mar 15 00:34:23 GMT 2013: 0.1.4b
- added PATH_ROOT to Origional files
Fri Mar 15 23:43:01 GMT 2013: 0.1.5b
- added ExpireDeadCacheFiles
- external JS defer sort before local
- clear cache button
Thu May  1 11:36:40 BST 2014: 0.1.6b
- 2.1 support
- sub-domain friendly
Thu May  1 14:52:57 BST 2014: 0.1.7b
- remove redundant in array $GlobalJS
Fri May  2 08:51:46 BST 2014: 0.1.8b
- preg_match format bug
Fri May  2 15:56:10 BST 2014: 0.1.9b
- cdn files load first
Thu Dec 11 10:18:12 GMT 2014: 0.1.10b
- base fix on complicated setups
Thu Dec  3 15:01:47 GMT 2015: 0.2.0b
- 2.2 compatibility
Mon Jan  9 22:13:03 GMT 2017: 0.2.1b
- 2.3 compatibility




js/consolidate.js
jQuery(document).ready(function($){
    $('body').on('click', 'a.ConsolidateRemove',function(e){
        e.preventDefault();
        $(this).parent('div.Chunk').remove();
    });
    $('body').on('keyup','div.ChunkGroupIndex div.Chunk',function(){
        if($(this).is(':last-child')){
            var ext = $($(this).children('.ConsolidateInput:eq(0)')).val().replace(/\s/g,'');
            var pattern = $($(this).children('.ConsolidateInput:eq(1)')).val().replace(/\s/g,'');
            if(ext!='' && pattern!=''){
                var clone = $(this).clone();
                clone.children('.ConsolidateInput:eq(1)').val('');
                $(this).after(clone);
            }
        }
    });
    
    $('body').on('click', 'div.ChunkGroupIndex div.Chunk .UpConsolidate',function(e){
        e.preventDefault();
        $(this).parents('div.Chunk').insertBefore($(this).parents('div.Chunk').prev());
    });
    
    $('body').on('click', 'div.ChunkGroupIndex div.Chunk .DownConsolidate',function(e){
        e.preventDefault();
        $(this).parents('div.Chunk').insertAfter($(this).parents('div.Chunk').next());
    });
});


CSS/Compressor.php
<?php namespace Consolidate;
/**
 * Class Minify_CSS_Compressor 
 * @package Minify
 */

/**
 * Compress CSS
 *
 * This is a heavy regex-based removal of whitespace, unnecessary
 * comments and tokens, and some CSS value minimization, where practical.
 * Many steps have been taken to avoid breaking comment-based hacks, 
 * including the ie5/mac filter (and its inversion), but expect tricky
 * hacks involving comment tokens in 'content' value strings to break
 * minimization badly. A test suite is available.
 * 
 * @package Minify
 * @author Stephen Clay <steve@mrclay.org>
 * @author http://code.google.com/u/1stvamp/ (Issue 64 patch)
 */
class Minify_CSS_Compressor {

    /**
     * Minify a CSS string
     * 
     * @param string $css
     * 
     * @param array $options (currently ignored)
     * 
     * @return string
     */
    public static function process($css, $options = array())
    {
        $obj = new Minify_CSS_Compressor($options);
        return $obj->_process($css);
    }
    
    /**
     * @var array
     */
    protected $_options = null;
    
    /**
     * Are we "in" a hack? I.e. are some browsers targetted until the next comment?
     *
     * @var bool
     */
    protected $_inHack = false;
    
    
    /**
     * Constructor
     * 
     * @param array $options (currently ignored)
     */
    private function __construct($options) {
        $this->_options = $options;
    }
    
    /**
     * Minify a CSS string
     * 
     * @param string $css
     * 
     * @return string
     */
    protected function _process($css)
    {
        $css = str_replace("\r\n", "\n", $css);
        
        // preserve empty comment after '>'
        // http://www.webdevout.net/css-hacks#in_css-selectors
        $css = preg_replace('@>/\\*\\s*\\*/@', '>/*keep*/', $css);
        
        // preserve empty comment between property and value
        // http://css-discuss.incutio.com/?page=BoxModelHack
        $css = preg_replace('@/\\*\\s*\\*/\\s*:@', '/*keep*/:', $css);
        $css = preg_replace('@:\\s*/\\*\\s*\\*/@', ':/*keep*/', $css);
        
        // apply callback to all valid comments (and strip out surrounding ws
        $css = preg_replace_callback('@\\s*/\\*([\\s\\S]*?)\\*/\\s*@'
            ,array($this, '_commentCB'), $css);

        // remove ws around { } and last semicolon in declaration block
        $css = preg_replace('/\\s*{\\s*/', '{', $css);
        $css = preg_replace('/;?\\s*}\\s*/', '}', $css);
        
        // remove ws surrounding semicolons
        $css = preg_replace('/\\s*;\\s*/', ';', $css);
        
        // remove ws around urls
        $css = preg_replace('/
                url\\(      # url(
                \\s*
                ([^\\)]+?)  # 1 = the URL (really just a bunch of non right parenthesis)
                \\s*
                \\)         # )
            /x', 'url($1)', $css);
        
        // remove ws between rules and colons
        $css = preg_replace('/
                \\s*
                ([{;])              # 1 = beginning of block or rule separator 
                \\s*
                ([\\*_]?[\\w\\-]+)  # 2 = property (and maybe IE filter)
                \\s*
                :
                \\s*
                (\\b|[#\'"-])        # 3 = first character of a value
            /x', '$1$2:$3', $css);
        
        // remove ws in selectors
        $css = preg_replace_callback('/
                (?:              # non-capture
                    \\s*
                    [^~>+,\\s]+  # selector part
                    \\s*
                    [,>+~]       # combinators
                )+
                \\s*
                [^~>+,\\s]+      # selector part
                {                # open declaration block
            /x'
            ,array($this, '_selectorsCB'), $css);
        
        // minimize hex colors
        $css = preg_replace('/([^=])#([a-f\\d])\\2([a-f\\d])\\3([a-f\\d])\\4([\\s;\\}])/i'
            , '$1#$2$3$4$5', $css);
        
        // remove spaces between font families
        $css = preg_replace_callback('/font-family:([^;}]+)([;}])/'
            ,array($this, '_fontFamilyCB'), $css);
        
        $css = preg_replace('/@import\\s+url/', '@import url', $css);
        
        // replace any ws involving newlines with a single newline
        $css = preg_replace('/[ \\t]*\\n+\\s*/', "\n", $css);
        
        // separate common descendent selectors w/ newlines (to limit line lengths)
        $css = preg_replace('/([\\w#\\.\\*]+)\\s+([\\w#\\.\\*]+){/', "$1\n$2{", $css);
        
        // Use newline after 1st numeric value (to limit line lengths).
        $css = preg_replace('/
            ((?:padding|margin|border|outline):\\d+(?:px|em)?) # 1 = prop : 1st numeric value
            \\s+
            /x'
            ,"$1\n", $css);
        
        // prevent triggering IE6 bug: http://www.crankygeek.com/ie6pebug/
        $css = preg_replace('/:first-l(etter|ine)\\{/', ':first-l$1 {', $css);
            
        return trim($css);
    }
    
    /**
     * Replace what looks like a set of selectors  
     *
     * @param array $m regex matches
     * 
     * @return string
     */
    protected function _selectorsCB($m)
    {
        // remove ws around the combinators
        return preg_replace('/\\s*([,>+~])\\s*/', '$1', $m[0]);
    }
    
    /**
     * Process a comment and return a replacement
     * 
     * @param array $m regex matches
     * 
     * @return string
     */
    protected function _commentCB($m)
    {
        $hasSurroundingWs = (trim($m[0]) !== $m[1]);
        $m = $m[1]; 
        // $m is the comment content w/o the surrounding tokens, 
        // but the return value will replace the entire comment.
        if ($m === 'keep') {
            return '/**/';
        }
        if ($m === '" "') {
            // component of http://tantek.com/CSS/Examples/midpass.html
            return '/*" "*/';
        }
        if (preg_match('@";\\}\\s*\\}/\\*\\s+@', $m)) {
            // component of http://tantek.com/CSS/Examples/midpass.html
            return '/*";}}/* */';
        }
        if ($this->_inHack) {
            // inversion: feeding only to one browser
            if (preg_match('@
                    ^/               # comment started like /*/
                    \\s*
                    (\\S[\\s\\S]+?)  # has at least some non-ws content
                    \\s*
                    /\\*             # ends like /*/ or /**/
                @x', $m, $n)) {
                // end hack mode after this comment, but preserve the hack and comment content
                $this->_inHack = false;
                return "/*/{$n[1]}/**/";
            }
        }
        if (substr($m, -1) === '\\') { // comment ends like \*/
            // begin hack mode and preserve hack
            $this->_inHack = true;
            return '/*\\*/';
        }
        if ($m !== '' && $m[0] === '/') { // comment looks like /*/ foo */
            // begin hack mode and preserve hack
            $this->_inHack = true;
            return '/*/*/';
        }
        if ($this->_inHack) {
            // a regular comment ends hack mode but should be preserved
            $this->_inHack = false;
            return '/**/';
        }
        // Issue 107: if there's any surrounding whitespace, it may be important, so 
        // replace the comment with a single space
        return $hasSurroundingWs // remove all other comments
            ? ' '
            : '';
    }
    
    /**
     * Process a font-family listing and return a replacement
     * 
     * @param array $m regex matches
     * 
     * @return string   
     */
    protected function _fontFamilyCB($m)
    {
        // Issue 210: must not eliminate WS between words in unquoted families
        $pieces = preg_split('/(\'[^\']+\'|"[^"]+")/', $m[1], null, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);
        $out = 'font-family:';
        while (null !== ($piece = array_shift($pieces))) {
            if ($piece[0] !== '"' && $piece[0] !== "'") {
                $piece = preg_replace('/\\s+/', ' ', $piece);
                $piece = preg_replace('/\\s?,\\s?/', ',', $piece);
            }
            $out .= $piece;
        }
        return $out . $m[2];
    }
}


views/consolidate.php
<?php if (!defined('APPLICATION')) exit(); ?>
<h1><?php echo $this->Data['Title']; ?></h1>
<?php
    echo $this->Form->open();
    echo $this->Form->errors();
?>
<style type="text/css">
input.SmallInput{
    width:100px;
}
input.ConsolidateInput{
    margin:0 5px;
}
.ConsolidateLabel{
    display:inline;
    float:none;
}

.ChunkGroupIndex{
    margin-top:30px;
}
</style>
<div class="Configuration">
   <div class="ConfigurationForm">
    <ul>
        <li>
            <?php
                echo $this->Form->label('Chunks'); 
                echo '<p> '.t('An ordered list of <a href="http://www.php.net/manual/en/pcre.pattern.php">PCRE</a> patterns used to divide resources in to consolidation chunks (applied first to last).').'</p>';
                echo '<p class="Warning"> '.t('Best left for those confident in what they are doing, some files need to be declared in a specific order to work properly. The defaults are recommended.').'</p>';
            ?>
        </li>
        <li>
            <?php
            $chunks = $this->data('Chunks');
            foreach($chunks as $chunkGroupIndex => $chunkGroup){
                echo '<div class="ChunkGroupIndex">';
                $chunkGroup['']='';
                foreach($chunkGroup as $chunk){
                    echo '<div class="Chunk">';
                    echo '<span class="ConsolidateLabel">'.t('File extension: <code>'.str_pad($chunkGroupIndex ? $chunkGroupIndex: '.js', 5, ' ').'</code> ').'</span>';
                    $this->Form->addHidden('ChunksExt[]',$chunkGroupIndex);
                    echo $this->Form->hidden('ChunksExt[]', array('value'=>$chunkGroupIndex, 'class'=>'SmallInput ConsolidateInput'));
                    echo '<span class="ConsolidateLabel">'.T('Match pattern:^').'</span>';
                    echo $this->Form->textBox('Chunks[]',array('value'=>$chunk, 'class'=>'InputBox ConsolidateInput'));
                    echo anchor('Remove','#',array('class'=>'SmallButton ConsolidateRemove'));
                    echo anchor(t('&uarr;'),'#',array('class'=>'UpConsolidate Button SmallButton'));
                    echo anchor(t('&darr;'),'#',array('class'=>'DownConsolidate Button SmallButton'));

                    echo '</div>';
                }
                echo '</div>';
            }
            ?>
        </li>
        <li>
            <?php 
            echo $this->Form->label('Defer JavasScript to Body foot'); 
            echo '<p class="Alert"> '.t('Warning! The following option to defer the JavasScript can easily break functionality. Use at your own risk. It will attempt to take declared JavaScript files and any inline JavaScript in the head, it can detect, and move it to the bottom. However there is no guarantee this will work as JavaScript can be added anywhere and in a magnitude of different ways, and these may depend on things that are ment to come before.').'</p>';
            ?>
        </li>
        <li>
            <?php
            echo '<span class="ConsolidateLabel">'.t('Defer JavaScript ').'</span>';
            echo $this->Form->dropDown('DeferJs', array(0=>'No', 1=>'Yes'), array('class'=>'SmallButton', 'value'=> c('Plugins.Consolidate.DeferJs'))); 
            ?>
        </li>
        <li>
            <?php echo $this->Form->button('Save Chunks', array('class'=>'SmallButton')); ?>
        </li>
    </ul>
    </div>
</div>
<?php
    echo $this->Form->close();
    echo $this->Form->open();
    echo $this->Form->errors();
?>
<div class="Configuration">
   <div class="ConfigurationForm">
    <ul>
        <li>
            <?php
                $this->Form->addHidden('ClearCache',1);
                echo $this->Form->hidden('ClearCache', array('value'=>1));
                echo $this->Form->label('Clear Cache');
                echo '<span class="ConsolidateLabel">'.t('This will wipe the cache, note users will have to explore the full site to get a complete cache. ').'</span>';
             ?>
        </li>
        </li>
            <?php echo $this->Form->button('Clear Cache',array('class'=>'SmallButton')); ?>
        </li>
    </ul>
    </div>
</div>
<?php
    echo $this->Form->close();


default.php
<?php
$PluginInfo['Consolidate'] = array(
   'Name' => 'Consolidate',
   'Description' => 'Combine resources (js, css) according to regex patterns, to have the ideal number of requests, and minimise bloat.',
   'Version' => '0.2.1b',
   'RequiredApplications' => array('Vanilla' => '2.3'),
   'Author' => "Paul Thomas",
   'AuthorEmail' => 'dt01pq_pt@yahoo.com',
   'AuthorUrl' => 'http://vanillaforums.org/profile/x00',
   'SettingsUrl' => '/dashboard/settings/consolidate',
   'MobileFriendy' => true,
   'License'=>"GNU GPL2"
);



define('CONSOLIDATE_ROOT', dirname(__FILE__));

include_once(CONSOLIDATE_ROOT.DS.'CSS'.DS.'UriRewriter.php');
include_once(CONSOLIDATE_ROOT.DS.'CSS'.DS.'Compressor.php');

class Consolidate extends Gdn_Plugin {

    protected $checked = false;

    protected $webRoot = '';

    protected $chunks = array();
    protected $chunkedFiles = array();
    protected $deferJs = array();
    protected $externalJs = array();
    protected $inlineJs = array();
    protected $inlineJsStrings = array();
    protected $cdns = array();
   
    protected function chunkedFiles($put = array(), $save = true){
        $token = 'chunked_files';
        $cacheFile = PATH_CACHE."/Consolidate/$token";
        //check cache
        if (file_exists($cacheFile)) {
            $put = array_merge($put,Gdn_Format::unserialize(file_get_contents($cacheFile)));
            $put = array_unique($put);
            if (!$save) {
                return $put;
            }
        }
        //save
        if ($save && !empty($put)) {
            if (!file_exists(dirname($cacheFile))) {
                mkdir(dirname($cacheFile), 0777, true);
            }
            file_put_contents($cacheFile, Gdn_Format::serialize($put));
        }
       
        return $put;
    }
   
    protected function getChunks(){
        $this->chunks = c('Plugins.Consolidate.Chunks',
        array(
                '.js' => array(
                    '/?js/library/jquery.js',
                    '/?js/global.js',
                    '/?js/.*',
                    '/?applications/.*',
                    '/?plugins/.*',
                    '/?themes/.*',
                    '.*'
                ),
                '.css' => array(
                    '/?applications/.*',
                    '/?resources/.*',
                    '/?plugins/.*',
                    '/?themes/.*',
                    '.*'
                )
            )
        );
      
        $this->chunkedFiles = $this->chunkedFiles(array());
    } 
   
    public function base_getAppSettingsMenuItems_handler($sender) {
        $menu = $sender->EventArguments['SideMenu'];
        $menu->addLink('Site Settings', t('Consolidate'), 'settings/consolidate', 'Garden.Settings.Manage');
    }
   
    public function settingsController_consolidate_create($sender, $args){
        $sender->permission('Garden.Settings.Manage');
        if ($sender->Form->isPostBack() != false) {
            $formValues = $sender->Form->formValues();
            if (val('ClearCache',$formValues)){
                $this->clearCache();
                redirect('settings/consolidate');
            }
            $chunkGroups = array();
            foreach ($formValues['Chunks'] As $chunkGroupIndex => $chunk) {
                
                if (!is_string($chunk) || preg_match('`^\s*$`',$chunk)) {
                    continue;
                }
                
                $ext = $formValues['ChunksExt'][$chunkGroupIndex];
                
                if (!is_string($ext)) {
                    continue;
                }
                
                if (!val($ext,$chunkGroups)) {
                    $chunkGroups[$ext] = array();
                }
                //in absense of escaping. 2.1
                $chunkGroups[$ext][] = trim(preg_replace('`([\'\$])`','\\\$1',$chunk));
            }
            
            foreach ($chunkGroups As $ext => $chunkGroup) {
                //ensure remainder (.*) is applied at end
                $i = array_search('.*',$chunkGroups[$ext]);
                if($i > -1){
                    unset($chunkGroups[$ext][$i]);
                }
                $chunkGroups[$ext][] = '.*';
            } 
            saveToConfig('Plugins.Consolidate.DeferJs',val('DeferJs', $formValues) ? true : false);
            if (!empty($chunkGroups)) {
                saveToConfig('Plugins.Consolidate.Chunks', $chunkGroups);
            }
            if (!empty($chunkGroups) || val('deferJs', $formValues)) {
                redirect('settings/consolidate');
            }
        }
        $sender->addSideMenu('settings/consolidate');
        $sender->setData('Title', t('Consolidate Settings'));
        $sender->addJsFile('consolidate.js', 'plugins/Consolidate');
        $this->getChunks();
        $sender->setData('Chunks', $this->chunks);
        $sender->Render($this->getView('consolidate.php'));
    }
   
    public function base_afterJscdns_handler($sender){
        $this->cdns = $sender->EventArguments['Cdns'];
    }
    
    public function headModule_beforeToString_handler($head) {
        if ($this->checked) {
            return;
        }
        
        $this->checked = true; 
       
        $this->webRoot = Gdn::request()->webRoot();
      
        $this->getChunks();
        $this->expireDeadCacheFiles();
      
        $tags = $head->tags();
        $cssToCache = array();
        $jsToCache = array(); 
        $externalJs = array();
        // Process all tags, finding JS & CSS files
        foreach ($tags as $index => $tag) {
            $isJs = val(HeadModule::TAG_KEY, $tag) == 'script';
            $isCss = val(HeadModule::TAG_KEY, $tag) == 'link' && val('rel', $tag) == 'stylesheet';
            if (!$isJs && !$isCss) {
                continue;
            }
            
            if ($isCss) {
                $href = val('href', $tag, '!');
            } else {
                $href = val('src', $tag, '!');
            
                if (isset($tag[HeadModule::CONTENT_KEY])) {
                    $href = '!';
                    unset($tags[$index]['src']);
                }
            
                if(c('Plugins.Consolidate.DeferJs') && $href=='!'){
                    $this->inlineJs[] = $tag;
                    unset($tags[$index]);
                    continue;
                }
                if(c('Plugins.Consolidate.DeferJs') && $href[0] != '/'){
                    $this->externalJs[] = $tag;
                    unset($tags[$index]);
                    continue;
                }
            }
        
            //ensure that cdn files are loaded first. 
            if (!empty($this->cdns) && in_array($href, $this->cdns)) {
                $tag['_sort']=intval($tag['_sort'])-300;
                $tags[$index] = $tag;
            }
            
        
             // Skip the rest if path doesn't start with a slash
            if ($href[0] != '/') {
                continue;
            }

            // Strip any querystring off the href.
            $hrefWithVersion = $href;
            $href = preg_replace('`\?.*`', '', $href);
         
             // Strip webRoot & extra slash from Href 
            if( $this->webRoot != '') {
                $href = preg_replace("`^/{$this->webRoot}/`U", '', $href);
            }

            // Skip the rest if the file doesn't exist
            $fixPath = ($href[0] != '/') ? '/' : ''; // Put that slash back to test for it in file structure
            $path = PATH_ROOT . $fixPath . $href;
            if (!file_exists($path)) {
                continue;
            }

            // Remove from the tag because consolidate is taking care of it.
            unset($tags[$index]);

            // Add the reference to the appropriate cache collection.
            if ($isCss) {
                $cssToCache[] = array('href'=>$href, 'fullhref'=>$hrefWithVersion);
            } elseif ($isJs) {
                $jsToCache[] = array('href'=>$href, 'fullhref'=>$hrefWithVersion);
            }
        }
      
        $head->tags($tags);
      
        $chunkedFilesTemp = $this->chunkedFiles;
        $chunkedCss = $this->chunk($cssToCache, '.css');
        foreach($chunkedCss As $cssChunkGroup => $cssChunk){
            $token = $this->consolidateFiles($cssChunk, '.css', $cssChunkGroup);
            $head->addCss("/cache/Consolidate/$token", 'screen', false);
        }
        $head = $this->seperateInlineScripts($head);
        Gdn::controller()->Assets['Head'] = $head;
        $chunkedJs = $this->chunk($jsToCache, '.js');
        foreach($chunkedJs As $jsChunkGroup => $jsChunk){
            $token = $this->consolidateFiles($jsChunk, '.js', $jsChunkGroup);
            if (!c('Plugins.Consolidate.DeferJs')) {
                $head->addScript("/cache/Consolidate/$token", 'text/javascript', (stripos($token,'_jquery_js')!==false) ? -100 : false);
            }else{
                $this->deferJs[]=$token;
            }
        }
      
        sort($this->chunkedFiles);
        sort($chunkedFilesTemp);
        if($this->chunkedFiles != $chunkedFilesTemp){
            $this->chunkedFiles($this->chunkedFiles, true);
        }
    }
   
    public function base_afterBody_handler($sender){
        if (!empty($this->deferJs)) {
            $head = new HeadModule();
            $sortLast = 0;

            foreach($this->externalJs As $externalJs){
                if(intval($externalJs['_sort'])>$sortLast){
                    $sortLast = intval($externalJs['_sort']);
                }
            }
            
            foreach($this->deferJs As $token){
                $sortLast++;
                $head->addScript("/cache/Consolidate/$token", 'text/javascript', (stripos($token, '_jquery_js')!==false) ? -100 : $sortLast);
            }
            $tags = $head->getTags();

            foreach($this->inlineJs as &$inlineJs) {
                if (stripos($inlineJs[HeadModule::CONTENT_KEY], 'gdn=window.gdn||{}')!==false) {
                    $inlineJs['_sort'] = -1000;
                }
            }
            $tags = array_merge($tags, $this->externalJs);
            $tags = array_merge($tags, $this->inlineJs);
            $head->tags($tags);
            $headParts = explode("\n", $head->ToString());
            $inScript = false;
            foreach($headParts As $headPart){
                if ($inScript || stripos($headPart, '<script')!==false) {
                    echo trim($headPart)."\n";
                    $inScript = stripos($headPart, '</script>')===false;
                }
            }
            foreach($this->inlineJsStrings As $inlineJsString){
                echo trim($inlineJsString)."\n";
            }
        }
    }
   
    protected function stripHTMLComments($matches){
        if(preg_match("`(<!--[\s]*\[if[^\]]*\]>[\s]*(-->)?)(.*?)((<!--)?[\s]*<!\[endif\][\s]*-->)`imsU", $matches[0])){
            return $matches[0];
        }
    }
   
    protected function escapeCommentedScript($matches){
        $quote = substr($matches[0], 0, 1);
        return preg_replace('`(scr)(ipt)`i', '\1'.$quote.'+'.$quote.'\2', $matches[0]);
    }
   
    protected function scriptSeperate($matches){
        $string = $matches[0];
        $this->inlineJsStrings[] = $string;
    }
   
    protected function stringsScriptSeperate($strings){
        $token = 'inline_'.md5(Gdn_Format::Serialize($strings));
        $cacheFile = PATH_CACHE."/Consolidate/$token";
        //check cache
        if (file_exists($cacheFile)) {
           $inline = Gdn_Format::unserialize(file_get_contents($cacheFile));
            if (val('Before', $inline)) {
                $strings = val('Before', $inline);
            }
            if (val('After',$inline)) {
                $this->inlineJsStrings = val('After', $inline);
            }
            if (!in_array($token, $this->chunkedFiles)) {
                $this->chunkedFiles[] = $token;
            }
           return $strings;
        }
        foreach($strings As &$string){ 
            //detect script
            if(stripos($string, '<script')!==false){
                //remove HTML comments
                $string = preg_replace_callback("`<!--(.*?)-->`imsU", array($this,'stripHTMLComments'), $string);    
                //escape quoted scripts
                $string = preg_replace_callback("`(?<!\\\\)'((.*?)<script[^>]*>(.*?)</script>(.*?))*?(?<!\\\\)'`imsU", array($this,'escapeCommentedScript'), $string);
                $string = preg_replace_callback("`(?<!\\\\)\"((.*?)<script[^>]*>(.*?)</script>(.*?))*?(?<!\\\\)\"`imsU", array($this,'escapeCommentedScript'), $string);
                //remove and save inline scripts (including conditional tags)
                $string = preg_replace_callback("`(<!--[\s]*\[if[^\]]*\]>[\s]*(-->)?)?<script[^>]*>(.*?)</script>((<!--)?[\s]*<!\[endif\][\s]*-->)?`imsU", array($this, 'scriptSeperate'), $string);
           }
        }
       
        $inline = array();
        if (!empty($strings)) {
            $inline['Before'] = $strings;
        }
        if (!empty($this->inlineJsStrings)) {
            $inline['After'] = $this->inlineJsStrings;
        }
        //cache
        if (!empty($inline)) {
            $inline = Gdn_Format::serialize($inline);
            if (!file_exists(dirname($cacheFile))) {
                mkdir(dirname($cacheFile), 0777, true);
            }
            file_put_contents($cacheFile, $inline);
            if(!in_array($token, $this->chunkedFiles)) {
                $this->chunkedFiles[] = $token;
            }
        }
       
        return $strings;
    }
   
    protected function seperateInlineScripts($head){
        if (c('Plugins.Consolidate.DeferJs')) {
            $strings = $head->getStrings();
            $refObject = new ReflectionObject($head);
            $refProperty = $refObject->getProperty('_Strings');
            $refProperty->setAccessible(true);
            $refProperty->setValue($head, array());
            $strings = $this->stringsScriptSeperate($strings);
            foreach($strings As $string) {
                $head->addString($string);
            }
       }
       
       return $head;
    }
  
   
    protected function chunk($files, $suffix) {
        $chunks = val(strtolower($suffix), $this->chunks);
        if (empty($chunks)) {
            return array(Gdn_Format::url('.*')=>$files);
        }
        $chunkedFiles = array();
        foreach($chunks As $chunk){
            foreach($files As $fileIndex => $file){
                if(preg_match('`^'.$chunk.'`', $file['href'])){
                    $niceKey = trim(Gdn_Format::Url(preg_replace('`[^a-z0-9]+`i', '_', $chunk)), '_').'_';
                    if(!val($niceKey, $chunkedFiles)) {
                        $chunkedFiles[$niceKey] = array();
                    }
                    $chunkedFiles[$niceKey][] = $file;
                    unset($files[$fileIndex]);
                }
            }
        }
        return  $chunkedFiles;
    }
   
   
    protected function consolidateFiles($files, $suffix, $prefix = '') { 
        $token = $prefix.md5($this->PluginInfo['Version'].serialize($files)).$suffix;
        $cacheFile = PATH_CACHE."/Consolidate/$token";
        if (in_array($token, $this->chunkedFiles)) {
            return $token;
        }
        if (!file_exists($cacheFile)) {
            $consolidateFile = '';
            $pathRootParts = explode(DS, PATH_ROOT);
            $webRootParts = explode(DS, $this->webRoot);
            $base = join(DS, array_slice($pathRootParts, 0, -count($webRootParts)));
            foreach ($files As $file) {
                $consolidateFile .= "/*\n";
                $consolidateFile .= "* Consolidate '{$file['fullhref']}'\n";
                $consolidateFile .= "*/\n\n";
                $originalFile = PATH_ROOT.DS.$file['href'];
                $fileStr = file_get_contents($originalFile);
                if ($fileStr && strtolower($suffix)=='.css') {
                    $fileStr = Consolidate\Minify_CSS_UriRewriter::rewrite($fileStr, dirname($base.DS.$this->webRoot.DS.$file['href']), $base);             
                    $fileStr = Consolidate\Minify_CSS_Compressor::process($fileStr);
                }
                $consolidateFile .= trim($fileStr);
                if ($fileStr && strtolower($suffix)=='.js') {
                    if (substr($consolidateFile, -1)!=';') {
                        $consolidateFile .= ";";
                    }
                }
                $consolidateFile .= "\n\n";
            }
            if (!file_exists(dirname($cacheFile))) {
                mkdir(dirname($cacheFile), 0777, true);
            }
            file_put_contents($cacheFile, $consolidateFile);
        }
        if (!in_array($token, $this->chunkedFiles)) {
            $this->chunkedFiles[] = $token;
        }
      
        return $token;
    }
   
    /**
    * Empty cache when disabling this plugin.
    */ 
    public function onDisable() { $this->ClearCache(); }
   
    /** 
    * Empty cache when enabling or disabling any other plugin, application, or theme.
    */
    public function settingsController_afterEnablePlugin_handler() { $this->ClearCache(); }
    public function settingsController_afterDisablePlugin_handler() { $this->ClearCache(); }
    public function settingsController_afterEnableApplication_handler() { $this->ClearCache(); }
    public function settingsController_afterDisableApplication_handler() { $this->ClearCache(); }
    public function settingsController_afterEnableTheme_handler() { $this->ClearCache(); }

    /**
    * Expire cache files.
    */
    private function expireDeadCacheFiles() {
        $expireLast = c('Plugins.Consolidate.ExpireLast');
        $expire = c('Plugins.Consolidate.ExpirePeriod', '1 Week');
        $expirePeriod = strtotime($expire);
        if (!$expireLast) {
            $expireLast = $expirePeriod;
        } 
        if ($expireLast<$expirePeriod) {
            return;
        }
        $chunkedFiles = $this->chunkedFiles;
        $files = glob(PATH_CACHE.'/Consolidate/*', GLOB_MARK);
        foreach ($files as $file) {
            $baseName = basename($file);
            if ($baseName=='chunked_files' || in_array($baseName, $chunkedFiles)) {
                continue;
            }
            if (filemtime($file)<$expirePeriod) {
                continue;
            }
            if (substr($file, -1) != '/') {
                unlink($file);
            }
        }
      
        saveToConfig('Plugins.Consolidate.ExpireLast', time());
    }
   
   
    /**
    * Empty cache.
    */
    private function clearCache() {
        $files = glob(PATH_CACHE.'/Consolidate/*', GLOB_MARK);
        foreach ($files as $file) {
            if (substr($file, -1) != '/') {
                unlink($file);
            }
        }
      
        $this->chunkedFiles(array(), true);
    }
   
}
